File: lib_common/src/main/java/com/soli/libCommon/net/upload/ProgressRequestBody.java
package com.soli.libCommon.net.upload;

import io.reactivex.Emitter;
import okhttp3.MediaType;
import okhttp3.RequestBody;
import okio.*;

import java.io.IOException;

/*
 * @author soli
 * @Time 2018/12/5 21:05
 */
class ProgressRequestBody extends RequestBody {

    private RequestBody mDelegate;
    private Emitter<UploadProgressInfo> mEmitter;
    private UploadProgressInfo mProgressInfo;
    private BufferedSink mBufferedSink;

    ProgressRequestBody(RequestBody delegate, Emitter<UploadProgressInfo> emitter,
                        UploadProgressInfo info) {
        mDelegate = delegate;
        mEmitter = emitter;
        mProgressInfo = info;
    }

    @Override
    public long contentLength() throws IOException {
        return mDelegate.contentLength();
    }

    @Override
    public MediaType contentType() {
        return mDelegate.contentType();
    }

    @Override
    public void writeTo(BufferedSink sink) throws IOException {
        if (sink instanceof Buffer) {
            // Log Interceptor
            mDelegate.writeTo(sink);
            return;
        }
        if (mBufferedSink == null) {
            mBufferedSink = Okio.buffer(wrapSink(sink));
        }
        mDelegate.writeTo(mBufferedSink);
        mBufferedSink.flush();
    }

    private Sink wrapSink(Sink sink) {
        return new ForwardingSink(sink) {

            @Override
            public void write(Buffer source, long byteCount) throws IOException {
                super.write(source, byteCount);
                if (mProgressInfo.total == 0) {
                    mProgressInfo.total = contentLength();
                }
                mProgressInfo.current += byteCount;
                mEmitter.onNext(mProgressInfo);
            }
        };
    }
}

File: lib_common/src/main/java/com/soli/libcommon/net/upload/ProgressRequestBody.kt
package com.soli.libcommon.net.upload

import com.soli.libcommon.net.download.FileProgressListener
import okhttp3.MediaType
import okhttp3.RequestBody
import okio.*
import java.io.IOException

/*
 * @author soli
 * @Time 2018/12/5 21:05
 */
class ProgressRequestBody(
    private val mDelegate: RequestBody,
    private val progressListener: FileProgressListener
) : RequestBody() {

    @Throws(IOException::class)
    override fun contentLength(): Long {
        return mDelegate.contentLength()
    }

    override fun contentType(): MediaType? {
        return mDelegate.contentType()
    }

    @Throws(IOException::class)
    override fun writeTo(sink: BufferedSink) {
        if (sink is Buffer) {
            // Log Interceptor
            mDelegate.writeTo(sink)
            return
        }
//        if (mBufferedSink == null) {
        val mBufferedSink = wrapSink(sink).buffer()
//        }
        mDelegate.writeTo(mBufferedSink)
        mBufferedSink.flush()
    }

    private fun wrapSink(sink: Sink): Sink {
        return object : ForwardingSink(sink) {
            private var progress = -1
            private var totalBytesSend = 0L
            @Throws(IOException::class)
            override fun write(source: Buffer, byteCount: Long) {
                super.write(source, byteCount)

                val contentLength = contentLength()

                val update = if (byteCount != -1L) byteCount else 0
                totalBytesSend += update

                val temp = ((100 * totalBytesSend) / contentLength).toInt()
                if (progress != temp) {
                    progress = temp
//                        RxJavaUtil.runOnUiThread {
                    progressListener.progress(
                        progress,
                        totalBytesSend,
                        update,
                        contentLength,
                        byteCount == -1L
                    )
//                        }
                }
            }
        }
    }
}
